Rewrite Internet Identity integration documentation. (#3922)
dfinity/portal: docs/developer-docs/identity/internet-identity/integrate-internet-identity.mdx

     this.greeting = `Your principal is: ${principal.toText()}`;
     this.#render();
   };
   ```

4. Lastly update the `#render` method to add the new "Who am I?" button and a click listener:
   ```js title="src/ii-integration-frontend/src/App.js"
   #render() {
     let body = html`
       <main>
         <img src="${logo}" alt="DFINITY logo" />
         <br />
         <br />
         <form action="#">
           <label for="name">Enter your name: &nbsp;</label>
           <input id="name" alt="Name" type="text" />
           <button type="submit">Click Me!</button>
         </form>
   // highlight-start
         <center>
           <button id="whoami">Who am I?</button>
         </center>
   // highlight-end
         <section id="greeting">${this.greeting}</section>
       </main>
     `;
     render(body, document.getElementById('root'));
     document
       .querySelector('form')
       .addEventListener('submit', this.#handleSubmit);
   // highlight-start
     document
       .querySelector('#whoami')
       .addEventListener('click', this.#handleWhoAmI);
   // highlight-end
   }
   ```

5. Then deploy the updated frontend:
   ```bash
   dfx deploy
   ```

> 6. Visit your deployed frontend, click the "Who am I?" button and it should return the anonymous principal:

Author: At this stage of the tutorial, the Internet Identity sign in has not need implemented yet, I'll rephrase your suggestion to omit Internet Identity.

   ```
   Your principal is: 2vxsx-fae
   ```

### Step 3: Implement Authentication

1. Install the required package
   ```bash
   npm install @dfinity/auth-client
   ```

   :::info

   Make sure that all **@dfinity/&lt;package&gt;** dependencies in **package.json** are the same version.

   :::

2. Add a method to the `App` class to get the correct Internet Identity url:
   ```js title="src/ii-integration-frontend/src/App.js"
   /**
    * Get the correct Internet Identity URL based on the current environment
    */
   #identityProvider = () => {
     if (process.env.DFX_NETWORK === "local") {
       return `http://${process.env.CANISTER_ID_INTERNET_IDENTITY}.localhost:4943`;
     } else if (process.env.DFX_NETWORK === "ic") {
       return `https://${process.env.CANISTER_ID_INTERNET_IDENTITY}.ic0.app`;
     } else {
       return `https://${process.env.CANISTER_ID_INTERNET_IDENTITY}.dfinity.network`;
     }
   };
   ```

   :::info

   In case you're developing on Safari, you need to change the value returned for a local development environment to `http://localhost:4943?canisterId=<canister_id>`.

   :::

> 3. Add a method to create an `AuthClient` instance and call it in the `App` class constructor:

Reviewer: ```suggestion
Add a method to create an `AuthClient` instance, then call it in the `App` class constructor:
```

   ```js title="src/ii-integration-frontend/src/App.js"
   constructor() {
     this.#render();
     await this.#createAuthClient();
   }

   /**
    * Replace backend actor identity with the identity from AuthClient,
    * additionally re-render to show the updated authentication status
    */
   #onIdentityUpdate = async () => {
     Actor.agentOf(ii_integration_backend).replaceIdentity(this.authClient.getIdentity());
     this.isAuthenticated = await this.authClient.isAuthenticated();
     this.#render();
   }

   /**
    * Create AuthClient, this loads an existing session if available
    */
   #createAuthClient = async () => {
     this.authClient = await AuthClient.create();
     await this.#onIdentityUpdate();
   }
   ```

> 4. Add methods to the `App` class to login and logout:

Reviewer: ```suggestion
Add methods to the `App` class to facilitate logging in and logging out:
```

   ```js title="src/ii-integration-frontend/src/App.js"
   /**
    * Login with AuthClient, this should be called without any delay in a click handler
    */
   #handleLogin = async () => {
     await new Promise((resolve, reject) => this.authClient.login({
       identityProvider: this.#identityProvider(),
       onSuccess: resolve,
       onError: reject
     }));
     await this.#onIdentityUpdate();
   }

   /**
    * Logout with AuthClient
    */
   #handleLogout = async () => {
     await this.authClient.logout();
     await this.#onIdentityUpdate();
   }
   ```

   :::warning

>    Make sure to always create a single AuthClient instance on page load and re-use it within your click handlers as shown in the previous and current step.

Reviewer: ```suggestion
   Make sure to always create a single `AuthClient` instance on page load and reuse it within your click handlers as shown in the previous and current steps.
```

   :::

> 5. Lastly update the `#render` method to add the new "Login" and "Logout buttons and click listeners:

Reviewer: ```suggestion
Lastly, update the `#render` method to add the new "Login" and "Logout" buttons and click listeners:
```

     let body = html`
       <main>
         <img src="${logo}" alt="DFINITY logo" />
         <br />
         <br />
         <form action="#">
           <label for="name">Enter your name: &nbsp;</label>
           <input id="name" alt="Name" type="text" />
           <button type="submit">Click Me!</button>
         </form>
         <center>
   // highlight-start
           ${this.isAuthenticated
             ? html`<button id="logout">Logout</button>`
             : html`<button id="login">Login</button>`}
   // highlight-end
           <button id="whoami">Who am I?</button>
         </center>
         <section id="greeting">${this.greeting}</section>
       </main>
     `;
     render(body, document.getElementById('root'));
     document
       .querySelector('form')
       .addEventListener('submit', this.#handleSubmit);
     document
       .querySelector('#whoami')
       .addEventListener('click', this.#handleWhoAmI);
   // highlight-start
     document
       .querySelector('#login')
       ?.addEventListener('click', this.#handleLogin);
     document
       .querySelector('#logout')
       ?.addEventListener('click', this.#handleLogout);
   // highlight-end
   }
   ```

> 6. Then deploy the updated frontend:

Reviewer: ```suggestion
Deploy the updated application:
```

   ```bash
   dfx deploy
   ```

> 7. Click the button to login, and you'll be redirected to the II frontend. Since you're running this locally, you will be using a local, non-production Internet Identity. To create one, follow the on-screen steps.

Reviewer: ```suggestion
Click the "Login" button. You'll be redirected to the Internet Identity frontend. Since you're running this locally, you will be using a local, non-production Internet Identity. To create one, follow the on-screen steps.
```

> 8. Click the "Who am I" button and it should return your Internet Identity principal:

Reviewer: ```suggestion
Click the "Who am I" button and it should return your Internet Identity principal:
```

   ```
   Your principal is: 5uylz-j7fcd-isj73-gp57f-xwwyy-po2ib-7iboa-fdkdv-nrsam-3bd3r-qqe
   ```

   :::info

>    The above principal is an example, the principal returned will be different based on the account and the environment where the dapp and Internet Identity are running.

Reviewer: ```suggestion
   The above principal is an example. The principal returned will be different based on the account and the environment where the application and Internet Identity are running.
```
